doc/loader.rst

```diff
@@ -5,128 +5,539 @@ MIG LOADER
-In normal scenarios, the MIG agent is deployed using some form of package, for
-example RPM or deb. When the agent needs to be updated, the older agent package
-is removed from the system and a new package is installed.
+The MIG loader is the component in MIG that can look after keeping agents up
+to date on systems. More specifically, this involves periodically checking if
+new agent code is available, making sure it's signed and safe to install, and
+handling automatic agent/configuration upgrades.
+
+In environments with a mature configuration management setup (such as via
+puppet or ansible) MIG loader may not be required. The configuration management
+system can periodically push updates out to required systems.
+
+In other environments, the loader can help keep things up to date. As an example
+this could be in cases where you want to keep MIG up to date on workstations that
+are not centrally managed. Even in cases with a configuration management system in
+place, you may want to use MIG loader to provide more fine-grained control over
+updates to agents.
+
+Here we describe how the loader works, and provide an example on how it is setup
+and used to provision and continuously upgrade a MIG environment.
+
+Differences between deploying with mig-loader and without
+---------------------------------------------------------
+Traditionally with MIG, if you wanted the agent installed on a set of systems,
+you would package the MIG agent up and install that agent package on the desired
+systems. If you wanted to upgrade the agent, you would need to upgrade the agent
+package installed on the systems.
+
+With mig-loader, instead of installing the agent on the systems you want to run
+the agent on, you would install only mig-loader. mig-loader is a small binary
+intended to be run from a periodic system such as cron. mig-loader will then
+look after fetching the agent and installing it if it does not exist on the system,
+and will look after upgrading the agent automatically if you want to publish new
+agent updates. The upgrades can be controlled by a MIG administrator through the
+MIG API and console tools.
+
+Key components
+--------------
+Several components within MIG are responsible for supporting loader related
+deployment. Some are dedicated to this, others have additional functionality
+that is used.
+
+mig-loader
+~~~~~~~~~~
+mig-loader is the component responsible for upgrading and managing the agent and
+associated agent configuration files on a system. It would run on a host you would
+like the agent deployed on. mig-loader looks after periodically checking if new
+agents are available by contacting the API, validating signatures on new agent
+manifests, and installing new agent code and configuration data when available. It
+also looks after management related actions such as restarting the agent if new
+code has been installed.
+
+mig-api
+~~~~~~~
+In the general sense, the MIG API is used by clients to send actions to agents and
```

Comment:
"the API is the control interface of MIG used by investigators to interact with the platform."